README.md

<!--- README.md is generated from README.Rmd. Please edit that file -->
Tiles for visualizing large spatial data
========================================

[![Project Status: WIP - Initial development is in progress, but there has not yet been a stable, usable release suitable for the public.](http://www.repostatus.org/badges/latest/wip.svg)](http://www.repostatus.org/#wip) [![Travis Build Status](https://img.shields.io/travis/jeffreyhanson/tiler/master.svg?label=build)](https://travis-ci.org/jeffreyhanson/tiler) [![Coverage Status](https://codecov.io/github/jeffreyhanson/tiler/coverage.svg?branch=master)](https://codecov.io/github/jeffreyhanson/tler?branch=master) [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/tiler)](https://CRAN.R-project.org/package=tiler)

The *tiler R* package can be used to render spatial data sets as [tiles](https://en.wikipedia.org/wiki/Tile_Map_Service). The advantage of rendering spatial data as tiles for use in web applications is that the tile format reduces the amount of data that users need to download to view the data.

Installation
------------

This package requires *Python* 2.7 to be installed on the system. Additionally, the *gdal* python library also needs to be installed. This can be achieved by running the following code in the terminal on Ubuntu 16.04.

``` bash
sudo apt-add-repository ppa:ubuntugis/ubuntugis-unstable
sudo apt-get update
sudo apt-get install gdal-bin python-gdal
```

Now, the *tiler R* package can be installed using the following *R* code:

``` r
if (!require(devtools))
  install.packages("devtools")
devtools::install_github("jeffreyhanson/tiler")
```

Example usage
-------------

Let's use this package to create tiles for visualizing a raster data set. Although the data set used in this example is rather small, the `tiles` function can be used to create tile data for large raster data sets. This means that *shiny* applications can include large raster data sets without users needing to wait a long time to view the data.

``` r
# load packages
library(leaflet)
library(tiler)

# load example data set
data(netherlands, package = "mapmisc")

# create leaflet map and visualize data using tiles
map <- leaflet() %>%
       setView(lng = 5.75560, lat = 50.94723, zoom = 12) %>%
       addProviderTiles("Esri.WorldImagery", group = "Basemap") %>%
       addTiles(tiles(nldElev, output_dir = tempdir(), zoom = "10-16"),
                group = "Elevation",
                options = tileOptions(minZoom = 5, maxZoom = 16,
                                      tms = TRUE)) %>%
       addLayersControl(baseGroups = "Basemap", overlayGroups = "Elevation",
                        options = layersControlOptions(collapsed = FALSE))

# render map
print(map)
```

Note that we set the argument `output_dir` to a temporary directory. For real-world applications, you would set this to a persistent directory so that the tiling process is only performed once.

Citation
--------

    Hanson JO (2017) tiler: Tiles for mapping large spatial data. R package version 0.0.0.4. https://github.com/jeffreyhanson/tiler.


NEWS.md
# tiler 0.0.0.4 (unreleased)

- Add _rgdal_ to Suggests.

# tiler 0.0.0.3 (unreleased)

- Remove _withr_ dependency.
- Add more comprehensive tests for `tiles`.

# tiler 0.0.0.2 (unreleased)

- Fix LICENSE.

# tiler 0.0.0.1 (unreleased)

- Initial package version.


tests/testthat.R
# set environmental variable
Sys.setenv("R_TESTS" = "")

# load packages
library(testthat)
library(tiler)

# run tests
test_check("tiler")


tests/testthat/test_tiles.R
context("tiles")

test_that("tiles", {
  skip_on_cran()
  # load data
  data(netherlands, package = "mapmisc")
  # create folder to save tiles
  d <- file.path(tempdir())
  dir.create(d, showWarnings = FALSE, recursive = TRUE)
  # make tiles
  expect_warning(p <- tiles(nldElev, d, zoom = "10-12", cache = FALSE))
  # tests
  expect_true(file.exists(d))
  expect_true(file.exists(file.path(d, "tiles")))
  expect_true(all(file.exists(file.path(d, "tiles", c(10:12)))))
  expect_true(file.exists(file.path(d, "tiles", "checksum.txt")))
  # cleanup
  unlink(d)
})

test_that("tiles (parallel)", {
  skip_on_cran()
  # load data
  data(netherlands, package = "mapmisc")
  # create folder to save tiles
  d <- file.path(tempdir(), "tiles2")
  dir.create(d, showWarnings = FALSE, recursive = TRUE)
  # make tiles
  expect_warning(p <- tiles(nldElev, d, zoom = "10-12", threads = 2,
                            cache = FALSE))
  # tests
  expect_true(file.exists(d))
  expect_true(file.exists(file.path(d, "tiles")))
  expect_true(all(file.exists(file.path(d, "tiles", c(10:12)))))
  expect_true(file.exists(file.path(d, "tiles", "checksum.txt")))
  # cleanup
  unlink(d)
})


tests/testthat/create_tiles.R
context("create_tiles")

test_that("make tiles", {
  # load data
  data(netherlands, package = "mapmisc")
  # create folder to save tiles
  d <- file.path(tempdir(), "tiles")
  # make tiles
  p <- create_tiles(nldEvel, d)
})


R/internal.R
#' Validate zoom data
#'
#' This function checks that zoom data are specified in a correct format.
#'
#' @param x \code{\link{character}} object.
#'
#' @return \code{\link{logical}} is it valid?
#'
#' @examples
#' # valid inputs
#' is_valid_zoom("10")
#' is_valid_zoom("1-10")
#'
#' # invalid inputs
#' is_valid_zoom("A")
#' is_valid_zoom("10.1")
#' is_valid_zoom("1-10.1")
#'
#' @noRd
is_valid_zoom <- function(x) {
  assertthat::assert_that(assertthat::is.string(x))
  if (grepl("-", x))
   x <- strsplit(x, "-")[[1]]
  for (i in seq_along(x)) {
    if (suppressWarnings(is.na(as.integer(x[i]))))
      return(FALSE)
    if (as.character(suppressWarnings(as.integer(x[i]))) != x[i])
      return(FALSE)
  }
  return(TRUE)
}

#' Color stack
#'
#' Create a \code{\link[raster]{RasterStack-class}} object containing color
#' values that correspond to data in a  \code{\link[raster]{RasterLayer-class}}
#' object.
#'
#' @param x \code{\link[raster]{RasterLayer-class}} object.
#'
#' @param colors The color palette (see \code{\link{colorNumeric}}) or function
#'   to use to color the raster values (hint: if providing a function, set
#'   \code{na.color} to \code{"#00000000"} to make \code{NA} areas transparent).
#'   Defaults to \code{"Spectral"}.
#'
#' @param opacity The base opacity of the raster, expressed from 0 to 1.
#'   Defaults to 1.
#'
#' @return \code{\link[raster]{RasterStack-class}} object with three bands. The
#'   bands correspond to red, green, blue, and alpha intensities.
#'
#' @examples
#' # load example data
#' data(netherlands, package = "mapmisc")
#'
#' # plot data
#' plot(nldElev, main = "elevation")
#'
#' # create color stack
#' nldElev2 <- color_stack(nldElev, "Spectral", 1)
#'
#' # plot color stack
#' plotRGB(nldElev2, main = "elevation")
#'
#' @noRd
color_stack <- function(x, colors = "Spectral", opacity = 1) {
  assertthat::assert_that(inherits(x, "RasterLayer"), raster::nlayers(x) == 1,
                          assertthat::is.string(colors) ||
                          inherits(colors, "function"),
                          assertthat::is.scalar(opacity),
                          opacity >= 0, opacity <= 1)
  ## convert colors to function if a character
  if (!is.function(colors))
    colors <- leaflet::colorNumeric(colors,
                                    domain = c(raster::cellStats(x, "min"),
                                               raster::cellStats(x, "max")),
                                    na.color = "#00000000", alpha = TRUE)
  ## apply color scheme
  # create empty raster data
  red <- suppressWarnings(raster::setValues(x, NA_real_))
  green <- suppressWarnings(raster::setValues(x, NA_real_))
  blue <- suppressWarnings(raster::setValues(x, NA_real_))
  alpha <- suppressWarnings(raster::setValues(x, NA_real_))
  # start writing data
  if (canProcessInMemory(x, 5)) {
    v <- grDevices::col2rgb(colors(raster::values(x)), alpha = TRUE)
    red <- raster::setValues(red, v[1, ])
    green <- raster::setValues(green, v[2, ])
    blue <- raster::setValues(blue, v[3, ])
    alpha <- raster::setValues(alpha, v[4, ])
  } else {
    f1 <- raster::rasterTmpFile()
    f2 <- raster::rasterTmpFile()
    f3 <- raster::rasterTmpFile()
    f4 <- raster::rasterTmpFile()
    red <- writeStart(red, f1)
    green <- writeStart(green, f2)
    blue <- writeStart(blue, f3)
    alpha <- writeStart(alpha, f4)

    # process data in chunks
    bs <- blockSize(x)
    for (i in seq_len(bs$n)) {
      v <- getValues(x, row = bs$row[i], nrows = bs$nrows[i] )
      v <- grDevices::col2rgb(colors(v), alpha = TRUE)
      red <- raster::writeValues(red, v[1, ], bs$row[i])
      green <- raster::writeValues(green, v[2, ], bs$row[i])
      blue <- raster::writeValues(blue, v[3, ], bs$row[i])
      alpha <- raster::writeValues(alpha, v[4, ], bs$row[i])
    }
    # finish writing files
    red <- raster::writeStop(red)
    green <- raster::writeStop(green)
    blue <- raster::writeStop(blue)
    alpha <- raster::writeStop(alpha)
    red <- raster::raster(f1)
    green <- raster::raster(f2)
    blue <- raster::raster(f3)
    alpha <- raster::raster(f4)
  }
  ## create raster stack
  out <- raster::stack(red, green, blue, alpha)
  names(out) <- c("red", "green", "blue", "alpha")
  return(out)
}


R/create_tiles.R
#' Create tiles
#'
#' Create tiles for a spatial dataset.
#'
#' @param x \code{\link[raster]{RasterLayer-class}} object.
#'
#' @param output_dir \code{\link{character}} directory in which to save tiles.
#'
#' @param zoom \code{character} Zoom levels to render. The levels can be
#'   specified as a single number (e.g \code{"10"}) or as a range
#'   (e.g. \code{"1-10"}).
#'
#' @param verbose \code{logical} Should information be printed during
#'   processing?
#'
#' @param threads \code{numeric} Number of threads to use for processing.
#'   Defaults to 1.
#'
#' @param cache \code{logical} if tiles already exist in argument to
#'    \code{output_dir} should the processing be skipped?
#'
#' @param ... not used.
#'
#' @details This function creates a tile layer, following the
#'   \href{https://wiki.osgeo.org/wiki/Tile_Map_Service_Specification}{Tile Map Service specification}, using the object in argument to \code{x} and saves it
#'   in the argument to \code{output_dir}.
#'
#' @seealso \code{\link[leaflet]{addTiles}}.
#'
#' @return \code{character} path containing the tile map template.
#'
#' @examples
#' # load leafleat
#' library(leaflet)
#'
#' # load data
#' data(netherlands, package = "mapmisc")
#'
#' # create directory to save tiles
#' d <- file.path(tempdir(), "tiles")
#' dir.create(d, showWarnings = FALSE, recursive = TRUE)
#'
#' # create tiles
#' p <- create_tiles(nldElev, d)
#'
#' \donttest{
#' # create leaflet map
#' m <- leaflet() %>%
#'      setView(lng = -71.0589, lat = 42.3601, zoom = 12) %>%
#'      addTiles(p)
#'
#' # render tiles in leaflet
#' print(m)
#' }
#' @export
create_tiles <- function(x, ...) UseMethod("create_tiles")

#' @rdname create_tiles
#' @method create_tiles default
#' @export
create_tiles.default <- function(x, ...)
  stop(paste0("Spatial data sets of type ", class(x)[1], "are not supported."))

#' @rdname create_tiles
#' @method create_tiles RasterLayer
#' @export
create_tiles.RasterLayer <- function(x, output_dir, zoom = "1-10",
                                     palette = "Spectral",
                                     verbose = FALSE, threads = 1,
                                     cache = TRUE) {
  ## validate input arguments
  assertthat::assert_that(inherits(x, "RasterLayer"),
                          assertthat::is.dir(output_dir),
                          assertthat::is.string(zoom),
                          assertthat::is.flag(verbose),
                          !is.na(verbose),
                          assertthat::is.count(threads),
                          threads <= parallel::detectCores(),
                          assertthat::is.flag(cache),
                          !is.na(cache))
  if (!is_valid_zoom(zoom))
    stop("argument to zoom is not correct format.")
  if (is.na(as.character(x@crs)))
    stop("argument to x must have a defined coordinate reference system.")
  if (!raster::compareCRS(x@crs, sp::CRS("+init=epsg:3857")))
    stop("argument to x must be projected to mercator (epsg:3857).")
  ## create tile data
  # create TLS template
  output_dir <- normalizePath(output_dir)
  tls <- paste0("file://", output_dir, "/tiles/{z}/{x}/{y}.png")
  dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)
  # save data to disk
  path <- file.path(tempdir(), "tiles.tif")
  x <- create_color_stack(x, palette)
  raster::writeRaster(x, path, NAflag =-9999, overwrite = FALSE)
  # check if tiles are already created
  key <- unname(tools::md5sum(path))
  if (cache)
    if (file.exists(file.path(output_dir, "tiles/checksum.txt")))
      if (readLines(file.path(output_dir, "tiles/checksum.txt")) == key)
        return(tls)
  # run tile generation script
  withr::with_dir(
    output_dir,
    system(paste0("python ",
      system.file("inst/python/gdal2tiles-multiprocess.py", package = "tiler"),
      " -z ", zoom, " -a -9999 --processes=", threads, ifelse(verbose,
      "--verbose", ""), " --leaflet \"", path, "\"")))
  # delete raster data from disk
  unlink(path)
  # store md5sum in tile folder
  writeLines(key, file.path(output_dir, "checksum.txt"))
  ## return result
  return(tls)
}


R/tiles.R
#' Tiles
#'
#' Create tiles for a spatial dataset.
#'
#' @param x \code{\link[raster]{RasterLayer-class}} object.
#'
#' @param output_dir \code{\link{character}} directory in which to save tiles.
#'
#' @param colors The color palette (see \code{\link{colorNumeric}}) or function
#'   to use to color the raster values (hint: if providing a function, set
#'   \code{na.color} to \code{"#00000000"} to make \code{NA} areas transparent).
#'   Defaults to \code{"Spectral"}.
#'
#' @param zoom \code{character} Zoom levels to render. The levels can be
#'   specified as a single number (e.g \code{"10"}) or as a range
#'   (e.g. \code{"1-10"}). Defaults to \code{"5-16"}.
#'
#' @param threads \code{numeric} number of threads to use for processing.
#'   Defaults to 1.
#'
#' @param project \code{logical} If argument to \code{x} is not already in the
#'   coordinate reference system expected by Leaflet (Mercator;
#'   \code{EPSG:3857}), should the data be automatically reprojected? If
#'   argument to \code{project} is \code{FALSE}, then the user will need
#'   to ensure that the data is in the correct coordinate reference system.
#'
#' @param verbose \code{logical} Should information be printed during
#'   processing?
#'
#' @param cache \code{logical} if tiles already exist in argument to
#'    \code{output_dir} should the processing be skipped?
#'
#' @param shiny \code{logical} are the tiles being served in a shiny
#'    applications?
#' @param ... not used.
#'
#' @details This function creates a tile layer, following the
#'   \href{https://wiki.osgeo.org/wiki/Tile_Map_Service_Specification}{Tile Map Service specification}, using the object in argument to \code{x} and saves it
#'   in the argument to \code{output_dir}.
#'
#' @seealso \code{\link[leaflet]{addTiles}}.
#'
#' @return \code{character} tile scheme.
#'
#' @examples
#' # load leafleat
#' library(leaflet)
#'
#' # load example data set
#' data(netherlands, package = "mapmisc")
#'
#' \donttest{
#' # create leaflet map and visualize data using tiles
#' map <- leaflet() %>%
#'        setView(lng = 5.75560, lat = 50.94723, zoom = 12) %>%
#'        addProviderTiles("Esri.WorldImagery", group = "Basemap") %>%
#'        addTiles(tiles(nldElev, output_dir = tempdir(), zoom = "10-16"),
#'                 group = "Elevation",
#'                 options = tileOptions(minZoom = 5, maxZoom = 16,
#'                                       tms = TRUE)) %>%
#'        addLayersControl(baseGroups = "Basemap", overlayGroups = "Elevation",
#'                         options = layersControlOptions(collapsed = FALSE))
#'
#' # render map
#' print(map)
#' }
#' @export
tiles <- function(x, ...) UseMethod("tiles")

#' @rdname tiles
#' @method tiles default
#' @export
tiles.default <- function(x, ...)
  stop(paste0("Spatial data sets of type ", class(x)[1], "are not supported."))

#' @rdname tiles
#' @method tiles RasterLayer
#' @export
tiles.RasterLayer <- function(x, output_dir, colors = "Spectral",
                              zoom = "5-16", threads = 1, project = TRUE,
                              verbose = FALSE, cache = TRUE, shiny = FALSE,
                              ...) {
  ## validate input arguments
  assertthat::assert_that(inherits(x, "RasterLayer"),
                          assertthat::is.dir(output_dir),
                          assertthat::is.string(zoom),
                          assertthat::is.count(threads),
                          threads <= parallel::detectCores(),
                          assertthat::is.flag(project),
                          assertthat::is.flag(verbose),
                          !is.na(verbose),
                          assertthat::is.flag(cache),
                          !is.na(cache),
                          assertthat::is.flag(shiny),
                          !is.na(shiny))
  if (!is_valid_zoom(zoom))
    stop("argument to zoom is not correct format.")
  ## create tile data
  # reproject data if needed
  if (project)
    x <- leaflet::projectRasterForLeaflet(x)
  # create TLS template
  output_dir <- normalizePath(output_dir)
  if (!shiny) {
    tls <- paste0("file://", output_dir, "/tiles/{z}/{x}/{y}.png")
  } else {
    shiny::addResourcePath("output", output_dir)
    tls <- "/output/tiles/{z}/{x}/{y}.png"
  }
  dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)
  # create color stack
  path <- file.path(output_dir, "tiles.tif")
  s <- color_stack(x, colors)
  # extract NODATA value
  nodata <- color_stack(raster::raster(matrix(c(NA_real_ , 1), ncol = 2,
                                                               nrow = 1)))[1]
  # save data to disk
  raster::writeRaster(s, path, overwrite = TRUE, datatype = "INT1U",
                      nodata = nodata[1])
  # check if tiles are already created
  key <- unname(tools::md5sum(path))
  if (cache)
    if (file.exists(file.path(output_dir, "tiles/checksum.txt")))
      if (readLines(file.path(output_dir, "tiles/checksum.txt")) == key)
        return(tls)
  # make tiles
  cmd <- paste0("python ",
                system.file("python", "gdal2tiles.py", package = "tiler"),
                " -z ", zoom, " -a ", paste(nodata, collapse = ","),
                " -w none --processes=", threads,
                ifelse(verbose, " --verbose", ""), " \"", path, "\" \"",
                output_dir, "/tiles\"")
  res <- system(cmd)
  if (!identical(res, 0L))
    stop("tile processing failed.")
  # delete raster data from disk
  unlink(path)
  # store md5sum in tile folder
  writeLines(key, file.path(output_dir, "tiles/checksum.txt"))
  ## return result
  return(tls)
}
